### PR TITLE
Refactor number emitting (match precision of ruby sass)

### PR DESCRIPTION
**Pull request by mgreter**
Fixes https://github.com/sass/libsass/issues/1153

### CODE REVIEW

File: inspect.cpp [hunk 1]
```diff
@@ -416,43 +416,91 @@ namespace Sass {
...
+      ss << fixed << value;
+    }
+
+    const char *it = ss.str().c_str();
+    while (*it && *it != '.' && *it != ',') ++ it;
+    const char *point = it;
+    while (*it && *it == '0') ++ it;
+    bool is_int = (*it == 0);
+    // take a shortcut for integers
+    if (is_int)
```
Comment (xzyfer) [inline inspect.cpp]: `bool is_int = d.find('.') == string::npos` ?

File: inspect.cpp [hunk 2]
```diff
@@ -416,43 +416,91 @@ namespace Sass {
...
+
+    // check if we got scientific notation in result
+    if (ss.str().find_first_of("e") != string::npos) {
+      ss.clear(); ss.str(string());
+      ss.precision(max(12, precision));
+      ss << fixed << value;
+    }
+
+    const char *it = ss.str().c_str();
+    while (*it && *it != '.' && *it != ',') ++ it;
```
Comment (xzyfer) [inline inspect.cpp]: I find this code really unintuitive and hard to grok. We should aim for expressive code first, and optimize later.

```
bool has_decimal = ss.find_first_of(".,");
```

File: inspect.cpp [hunk 3]
```diff
@@ -416,43 +416,91 @@ namespace Sass {
...
-    if (d == "-0") d.erase(0, 1);
-    // use fractional output if we had
-    // a value before it got truncated
-    if (d == "0" && nonzero) d = "0.0";
-    // if the precision is 0 sass cast
-    // casts to a float with precision 1
-    if (ctx->precision == 0) d += ".0";
-    // append number and unit
-    append_token(d + n->unit(), n);
+      error(res + " isn't a valid CSS value.", n->pstate());
```
Comment (xzyfer) [inline inspect.cpp]: Should this be in the output phase? Surely we know if a number in valid in eval?

Comment (mgreter) [inline inspect.cpp]: A number can eval to something that is not valid, like:

```
$a: 2px * 2px; // ->invalid
div { margin: $a / 2px; } // ->valid
```

File: inspect.cpp (hunk 2 above)
Comment (mgreter) [inline inspect.cpp]: Unoftunately every bit (even the number 12) seems to be exactly needed this way.  But I guess this one loop can be rewritten with find_first_of ...

**Comment by mgreter** [comment]
Refactored the code and incorporated the feedback from @xzyfer. Made the code much more explicit. I think I cannot really improve much on it anymore. Hope CI build passes this times. IMHO appveyor sometimes fails for one build (with random errors it seems) ...

**Comment by xzyfer** [comment]
:+1:

**Pull request merged**